Add MCUboot support for RV32M1_Vega

lylezhu2014 opened the pull request:
1. Enable flash driver for RV32M1 SOC.
2. Add MCUboot support for RV32M1_Vega

lylezhu2012 committed:
driver: flash: add flash driver for the RV32M1 SOC

Add driver and device tree for the flash controller for the RV32M1 SOC

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

MaureenHelm commented on dts/riscv32/rv32m1.dtsi:
```diff
@@ -289,17 +289,17 @@
 			#size-cells = <1>;
 
-			m4_flash: flash@0 {
+			flash0: flash@0 {
```
Why are you changing this? Even though we're not using the cortex m cores, the hardware reference manual refers to these regions as m4 and m0 flashes.

lylezhu2014 replied:
I removed it temporarily. And then I will try to raise a requirement to update the RM to make the flash region name easy to understand.

dleach02 replied:
That was my doing @MaureenHelm . In my version of this change I didn't adopt the cortex naming and discussed with Lyle about changing the name.

MaureenHelm commented on boards/riscv32/rv32m1_vega/rv32m1_vega_ri5cy_defconfig:
```diff
@@ -12,3 +12,4 @@ CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
 CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_MULTI_LEVEL_INTERRUPTS=y
+CONFIG_FLASH=y
```
This should not be in the default board config; it should get set in an application config.

lylezhu2014 replied:
Done

MaureenHelm commented on soc/riscv32/openisa_rv32m1/dts_fixup.h:
```diff
@@ -0,0 +1,18 @@
+#define DT_FLASH_DEV_BASE_ADDRESS		\
+				DT_OPENISA_RV32M1_FTFE_40023000_BASE_ADDRESS
+#define DT_FLASH_DEV_NAME				\
+				DT_OPENISA_RV32M1_FTFE_40023000_LABEL
```
Long lines are ok for dts fixups.

lylezhu2014 replied:
OK, updated.

lylezhu2012 committed:
boards: riscv32: rv32m1_vega: enable MCUboot for ri5cy core

1. Add flash partitions.
2. Add macro DT_START_UP_ENTRY_OFFSET. The entry of the RV32M1 is
   not the start of the vector table. Add the macro to inform the
   entry offset.
3. Update linker file to support MCUboot
   a. For normal cases (CONFIG_BOOTLOADER_MCUBOOT is cleared), the
   vector table is located last 256bytes of the flash.
   b. If CONFIG_BOOTLOADER_MCUBOOT is set, the vector table is located
   after the image header of MCUboot.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
dts: riscv32: update flash controller compatibility property

- Change flash controller compatibility property from nxp,kinetis-ftfe
to openisa,rv32m1-ftfe.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2014 commented on dts/bindings/flash_controller/openisa,rv32m1-ftfe.yaml:
```diff
@@ -0,0 +1,15 @@
+---
+title: NXP Kinetis Flash Memory Module (FTFE)
```
Done

lylezhu2014 commented on dts/bindings/flash_controller/openisa,rv32m1-ftfe.yaml:
```diff
@@ -0,0 +1,15 @@
+version: 0.1
+
+description: >
+    This binding gives for the NXP Kinetis Flash Memory Module E (FTFE)
```
Done

dleach02 commented on dts/riscv32/rv32m1.dtsi:
```diff
@@ -282,7 +282,7 @@
 
 		flash-controller@40023000 {
-			compatible = "nxp,kinetis-ftfe";
+			compatible = "openisa,rv32m1-ftfe";
```
I would follow the normal naming with:

fctrl0: flash-controller@40023000

lylezhu2014 replied:
This issue is not included by current pull request. It is better to create a new one for this issue.

MaureenHelm reviewed (approved):
Please rebase to fix the merge conflict
